### PR TITLE
Fix API used to retieve projects

### LINKED ISSUE #145: Only Sonarqube admins were able to generate the CNES Reports
### Expected behavior
Users who have access to only certain projects & are not administrators in Sonarqube are unable to generate the CNES report.
When clicking the CNES Report, I am seeing a **403** for **/api/projects/search** 
### Actual behavior
Anyone with access to certain projects should be able to download the reports

### Steps to reproduce behavior
Create a user and provide sonar-user access instead of sonar-administrators and try downloading a CNES report and you will see this error.

### Detection version
3.2.2

### LINKED ISSUE #165: Plugin is not working for non-admin users
#### Describe the bug

The plugin is working when I log in to SonarQube as an admin but not for non-admin users, it keeps on loading on a blank screen

#### To reproduce

> _Steps to reproduce the behavior._

1. Login as non-admin
2. Click on 'more>CNES'
3. See error

#### Expected behavior

> _A clear and concise description of what you expected to happen._

#### Screenshots & log

> _If applicable, add screenshots and logs to help explain your problem._ **Do not forget to obfuscate them if they contain sensitive data.**.
image

#### User environment 

> _Please complete the following information._

* OS: Windows 10
* Project version: [1.0.0]

### LINKED ISSUE #171: Get infinite number of projects
#### Is your feature request related to a problem? Please describe.

With the new used API "/api/components/search", we can print maximum 500 projects per page.
We need to concatenate this API if there are more than 500 projects to print in the drop-down list.

### PR DESCRIPTION
**Pull request by diegorodriguez31**
# Fixed issues
* Fix #165 
* Fix #145 
* Fix #171 

# Proposed changes
* The non-admin users weren't able to load the cnes-report for their projects. Now, they can.
* The maximum number of projects displayed was 500. Now, there is no limit.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,41 +1,41 @@
...
-            "@babel/env",
-            "@babel/react"
-        ],
-        "plugins": [
-            "@babel/plugin-proposal-class-properties"
-        ]
-    },
-    "dependencies": {}
-}
+{
```
Comment (Sancretor) [inline package.json:1]: From what I see, this file was not modified at all...
Maybe we should revert the change in Git, in case there are some hidden characters.
What do you think ?

File: src/main/js/common/api.js [hunk 2]
```diff
@@ -4,42 +4,35 @@
...
-export function getProjectsList() {
-  return getSonarVersion().then(version => {
-    // Depending on SonarQube version, the API is not the same to retrieve projects list
-    if (version >= 8.0) {
-      return getProjectsList8();
-    } else {
-      return getProjectsList79();
+//Functions used to get all user projects
+export function getProjectsList(){
+  let elementByPage = 500;
```
Comment (Sancretor) [inline src/main/js/common/api.js:16]: If this variable will never be modified, you should replace "let" by "const".

File: src/main/js/common/api.js [hunk 3]
```diff
@@ -4,42 +4,35 @@
...
+  let allPromises = [];
+  let finalProjectsList = [];
+  return getJSON("/api/components/search", {"qualifiers": "TRK", "ps":elementByPage}).then(response => {
+    let nbProjects = response.paging.total;
```
Comment (Sancretor) [inline src/main/js/common/api.js:20]: If this variable will never be modified, you should replace "let" by "const".

File: src/main/js/common/api.js [hunk 4]
```diff
@@ -4,42 +4,35 @@
...
+    for(let i = 1; i <= nbPages; i++){
+      allPromises.push(getJSON("/api/components/search", {"qualifiers": "TRK", "ps":elementByPage, "p":i}).then(response => {
+        return response.components;
+      }));
     }
+    return Promise.all(allPromises).then((results) => {
+      let projects = [];
+      results.forEach((result) => {
+        projects = projects.concat(result);
+        console.log(projects);
```
Comment (Sancretor) [inline src/main/js/common/api.js:31]: You should remove this debug command, as it will write data on every client's browser console.

File: src/main/js/common/api.js [hunk 5]
```diff
@@ -4,42 +4,35 @@
...
+      })
+      finalProjectsList = projects;
+      return finalProjectsList;
```
Comment (Sancretor) [inline src/main/js/common/api.js:34]: Here, I think you could directly return "projects" instead of using another variable.

File: src/main/js/common/api.js [hunk 6]
```diff
@@ -4,42 +4,35 @@
...
-    // Depending on SonarQube version, the API is not the same to retrieve projects list
-    if (version >= 8.0) {
-      return getProjectsList8();
-    } else {
-      return getProjectsList79();
+//Functions used to get all user projects
+export function getProjectsList(){
+  let elementByPage = 500;
+  let allPromises = [];
+  let finalProjectsList = [];
```
Comment (Sancretor) [inline src/main/js/common/api.js:18]: See my comment below regarding that same variable, bu it may not be useful anymore.

File: src/main/js/common/api.js [hunk 7]
```diff
@@ -4,42 +4,35 @@
...
+  return getJSON("/api/components/search", {"qualifiers": "TRK", "ps":elementByPage}).then(response => {
+    let nbProjects = response.paging.total;
+    let nbPages = Math.ceil(nbProjects/elementByPage);
+    for(let i = 1; i <= nbPages; i++){
+      allPromises.push(getJSON("/api/components/search", {"qualifiers": "TRK", "ps":elementByPage, "p":i}).then(response => {
+        return response.components;
+      }));
     }
+    return Promise.all(allPromises).then((results) => {
```
Comment (Sancretor) [inline src/main/js/common/api.js:27]: In order to avoid having too much nesting "then" methods, I think you should move this one and put it right after the previous "then".
So that you would have "getJson" -> then (returning Promise.all) -> then (to return concatenated results).
What do you think ?

File: yarn-error.log [hunk 8]
```diff
@@ -0,0 +1,6241 @@
+Arguments:
```
Comment (Sancretor) [inline yarn-error.log:1]: This file contains error logs, so it should not be added to the git projet.
Would you please remove it ?

**Review by Sancretor — CHANGES_REQUESTED** [review]
Could you please consider the comments I made, before we can merge this PR ?
Thanks.

File: src/main/js/common/api.js (hunk 7 above)
Comment (diegorodriguez31) [reply src/main/js/common/api.js:27]: I don't really understand your idea.
I prefer not to touch this promise because I do not control its scope at 100% for the moment.

**Review by Sancretor — DISMISSED** [review]

**Review by Sancretor — APPROVED** [review]

File: src/main/js/common/api.js [hunk 9]
```diff
@@ -4,43 +4,39 @@
...
-  });
+    //Wait until all promises are done
+    return Promise.all(allPromises);
+  }).then((results) => {
+    //Concatenate an array : x * 500 (x = number of pages)
+      let projects = [];
+      results.forEach((result) => {
+        projects = projects.concat(result);
+      })
+      return projects;
```
Comment (begarco) [inline src/main/js/common/api.js:38]: @diegorodriguez31 I tested it, and it works, congrats ;)
But we could improve the UX by sorting the final list, no?

**Review by begarco — CHANGES_REQUESTED** [review]
Congrats @diegorodriguez31 👍 

I tested it and it works great.

Can you just:
- fix the small defects
- have a look on how we could sort this list
- answer my question on dependencies

File: src/main/js/common/api.js (hunk 9 above)
Comment (Sancretor) [reply src/main/js/common/api.js:38]: I agree, it would be better.

**Review by begarco — APPROVED** [review]
